_posts/2017-11-21-Installing-Jenkins-with-PowerShell-DSC.md

```diff
@@ -0,0 +1,745 @@
+cChocoPackageInstaller installVisualStudioWebWorkload
+{
+	Name = "visualstudio2017-workload-netweb"
+	Params = "--includeOptional"
+	DependsOn = "[cChocoInstaller]installChoco","[cChocoPackageInstaller]installVisualStudio"
+}
+
+# Install NuGet
+File installNuget 
+{
+	DestinationPath = "C:\tools\nuget\nuget.exe"
+	SourcePath = (Join-Path $InstallConfDirectory "nuget.exe")
+	Ensure = "Present"
+	Type = "File"
+	Checksum = "modifiedDate"
+	Force = $true
+	MatchSource = $true
+}
+```
+
+I think that the above script is quite self-explanatory although I made it shorter for easier reading. The full script can be found [here](https://github.com/solita/powershell-dsc-jenkins/blob/master/jenkins_dsc.ps1) We have bunch of resources, most of them depends on something earlier. Afterwards we want to have all those resources in place. Only odd thing is in the installVisualStudioWebWorkload where I add includeOptional parameter for the choco. This is to get F# installed on the target machine as well. It does not come by default. If you need to dig on to those additional options in nuget packages you should head to chocolateys web page and investigate the packages you need to figure out if there is any customization options.
+
+## Installing custom made stuff
+
+We have something in our GitHub repositories that we would like to use with Jenkins. Those are marvelous libraries jmeter-perfotrator and powershell-zap. We can fetch the git repository directly into Jenkins server and whatever what is there with the help of just installed git client. Here is an example resource for that.
+
+```powershell
+# Install powershell-zap module 		
+Script installPowershellZap 
+{
+	GetScript = {
+		return @{ Result = gci "C:\tools\powershell-zap-master" }
+	}
+	SetScript = {
+		mkdir "C:\tools\powershell-zap-master"
+		$gitexe = "${ENV:ProgramFiles}\Git\cmd\git.exe"
+		$arguments = 'clone https://github.com/solita/powershell-zap.git "C:\tools\powershell-zap-master"'
+		$null = start-process $gitexe $arguments 
+	}
+	TestScript = {
+		Return (Test-Path "C:\tools\powershell-zap-master")
+	}
+	DependsOn = "[cChocoPackageInstaller]installJdk8","[cChocoPackageInstaller]installGit" 
+}
+```
+
+## The ugly
+
+Only big let down that I have had with DSC is manipulating environments PATH variable. You can do it really easily but only once. 
+
+```powershell
+Environment setVS2017ToolsPath 
+{
+	Name = 'PATH'
+	Ensure = 'Present'
+	Path = $true
+	DependsOn = "[cChocoPackageInstaller]installVisualStudio"
+	Value = "${ENV:ProgramFiles(x86)}\Microsoft Visual Studio\2017\Professional\Common7\IDE"
+}
+```
+
+After you have done that once and try to manipulate the exactly same named environment variable in the same configuration you will get error. Same applies also creating and deleting same file in the same configuration. Instead you can workaround with having multiple configurations or by using script resources in DSC. I have used script resources for this and well it is really ugly. 
```

Comment:
get *an* error

Reply:
*The* same applies also *for* creating... *the* same file

Instead,

*work around it by* tai *do a workaround by*